src/runtime/virtcontainers/utils/utils.go

```diff
@@ -312,3 +315,75 @@ func ConvertNetlinkFamily(netlinkFamily int32) pbTypes.IPFamily {
 		return pbTypes.IPFamily_v4
 	}
 }
+
+// WaitLocalProcess waits for the specified process for up to timeoutSecs seconds.
+//
+// Notes:
+//
+// - If the initial signal is zero, the specified process is assumed to be
+//   attempting to stop itself.
+// - If the initial signal is not zero, it will be sent to the process before
+//   checking if it is running.
+// - If the process has not ended after the timeout value, it will be forcibly killed.
+func WaitLocalProcess(pid int, timeoutSecs uint, initialSignal syscall.Signal, logger *logrus.Entry) error {
+	var err error
+
+	// Don't support process groups
+	if pid <= 0 {
+		return errors.New("can only wait for a single process")
+	}
+
+	if initialSignal != syscall.Signal(0) {
+		if err = syscall.Kill(pid, initialSignal); err != nil {
+			if err == syscall.ESRCH {
+				return nil
+			}
+
+			return fmt.Errorf("Failed to send initial signal %v to process %v: %v", initialSignal, pid, err)
+		}
+	}
+
+	pidRunning := true
+
+	// Wait for the VM process to terminate
+	startTime := time.Now()
+
+	for {
+		var _status syscall.WaitStatus
+		var _rusage syscall.Rusage
+		var waitedPid int
+
+		// "A watched pot never boils" and an unwaited-for process never appears to die!
+		waitedPid, err = syscall.Wait4(pid, &_status, syscall.WNOHANG, &_rusage)
+
+		if waitedPid == pid && err == nil {
+			pidRunning = false
+			break
+		}
+
+		if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
+			pidRunning = false
+			break
+		}
+
+		if time.Since(startTime).Seconds() >= float64(timeoutSecs) {
```

Comment:
uhmmm I'm afraid :worried: - this looks prone to hacks, what happen if system's time is the reset to `startTime`?  maybe we shouldn't rely on the system's time .. ?  can you use a channel timeout ?

Reply:
Hi @devimc - this is simply a rework of what we currently do... in three different places ;) Maybe a follow-on PR could be used to address improving the logic.

Reply:
I guess I could keep the top commit (to show the code was just moved), and then add another commit to rework using a channel...

Reply:
:+1: ok, sounds good to me, thanks

Reply:
@devimc - branch updated.